### PR TITLE
refresh token rotation

### LINKED ISSUE #519: Refresh tokens should only work once and return a new refresh token
Also known as [_"Refresh Token Rotation"_](https://tools.ietf.org/html/rfc6819#section-5.2.2.3)

Currently a refresh token can be redeemed for an access/ID token as many times at the hold of that refresh token likes. E.g.

```json
{
  "access_token":"2YotnFZFEjr1zCsicMWpAA",
  "token_type":"example",
  "expires_in":3600,
  "id_token": "..."
}
```

Instead, Dex should also return a new refresh token every time one is claimed.

```json
{
  "access_token":"2YotnFZFEjr1zCsicMWpAA",
  "token_type":"example",
  "expires_in":3600,
  "refresh_token":"tGzv3JOkF0XG5Qx2TlKWIA",
  "id_token": "..."
}
```

This immediately invalidates the current refresh token, and the client must use the new one. The primary motivation is detecting that a refresh token has be stolen. Summed up well by [rfc 6819](https://tools.ietf.org/html/rfc6819#section-5.2.2.3)

> Refresh token rotation is intended to automatically detect and
> prevent attempts to use the same refresh token in parallel from
> different apps/devices.  This happens if a token gets stolen from the
> client and is subsequently used by both the attacker and the
> legitimate client.

### PR DESCRIPTION
**Pull request by rsoletob**
Added refresh token rotation.

Fixes #519

### CODE REVIEW

File: server/server_test.go [hunk 1]
```diff
@@ -576,23 +576,25 @@ func TestServerRefreshToken(t *testing.T) {
...
-		token         string
-		clientID      string // The client that associates with the token.
-		creds         oidc.ClientCredentials
-		signer        jose.Signer
-		createScopes  []string
-		refreshScopes []string
-		expectedAud   []string
-		err           error
+		token                string
+		expectedRefreshToken string
```
Comment (ericchiang) [inline server/server_test.go]: How are you coming up with these values? Can you implement a helper function to do this?

File: db/refresh.go [hunk 2]
```diff
@@ -188,6 +188,66 @@ func (r *refreshTokenRepo) Revoke(userID, token string) error {
...
+		return "", refresh.ErrorInvalidToken
+	}
+
+	// Renew refresh token
+	tokenPayload, err = r.tokenGenerator.Generate()
+	if err != nil {
+		return "", err
+	}
+
+	payloadHash, err := bcrypt.GenerateFromPassword(tokenPayload, bcrypt.DefaultCost)
```
Comment (ericchiang) [inline db/refresh.go]: I forget what the point of this payload hash is. Might be beyond the scope of this PR, but we need to add a comment or remove this logic.

File: db/refresh.go [hunk 3]
```diff
@@ -188,6 +188,66 @@ func (r *refreshTokenRepo) Revoke(userID, token string) error {
 	return tx.Commit()
 }
 
+func (r *refreshTokenRepo) RenewRefreshToken(clientID, userID, oldToken string) (newRefreshToken string, err error) {
```
Comment (ericchiang) [inline db/refresh.go]: A lot of this logic looks identical to Create and Verify. Can we figure out how to have these share code? E.g. renewing the token should just be "Verify -> Delete -> Create"

File: server/server.go [hunk 4]
```diff
@@ -560,34 +560,34 @@ func (s *Server) CodeToken(creds oidc.ClientCredentials, sessionKey string) (*jo
 	return jwt, refreshToken, nil
 }
 
-func (s *Server) RefreshToken(creds oidc.ClientCredentials, scopes scope.Scopes, token string) (*jose.JWT, error) {
+func (s *Server) RefreshToken(creds oidc.ClientCredentials, scopes scope.Scopes, token string) (*jose.JWT, string, error) {
```
Comment (ericchiang) [inline server/server.go]: Are the raw Claims ever used or do they immediately call `Encode()`?

**Comment by ericchiang** [comment]
A few comments, but overall looks good. Thanks!

File: db/refresh.go (hunk 2 above)
Comment (rsoletob) [inline db/refresh.go]: I'm going to add a comment. It checks if the hash of payload from received refresh token are equal to the stored in database.

File: server/server.go (hunk 4 above)
Comment (rsoletob) [inline server/server.go]: Raw claims are encoded immediately [here](https://github.com/coreos/dex/blob/0e94e7625551a0e2c455de6d2568489e02e00f98/server/http.go#L534). But the interface [OIDCServer](https://github.com/coreos/dex/blob/0e94e7625551a0e2c455de6d2568489e02e00f98/server/server.go#L46) uses jose.JWT in more places. Do you prefer to change it?

**Comment by rsoletob** [comment]
Thanks for your comments. I see an user could have some refresh tokens for the same client and there is a [TODO](https://github.com/coreos/dex/blob/0e94e7625551a0e2c455de6d2568489e02e00f98/db/refresh.go#L101).  Is it correct? I can change it here or in another PR if you think it isn't a valid feature.

**Comment by ericchiang** [comment]
Probably best for a different PR :)

File: server/server.go (hunk 4 above)
Comment (ericchiang) [inline server/server.go]: No, this is fine. Again, it's something we can clean up later.

**Comment by rsoletob** [comment]
Ok @ericchiang, revamped

File: db/refresh.go [hunk 5]
```diff
@@ -91,47 +91,108 @@ func NewRefreshTokenRepoWithGenerator(dbm *gorp.DbMap, gen refresh.RefreshTokenG
...
 	}
 
-	// TODO(yifan): Check the number of tokens given to the client-user pair.
-	tokenPayload, err := r.tokenGenerator.Generate()
+	return tx.Commit()
+}
+
+func (r *refreshTokenRepo) RenewRefreshToken(clientID, userID, oldToken string) (newRefreshToken string, err error) {
+	// Verify
+	userID, connectorID, scopes, err := r.verify(nil, clientID, oldToken)
```
Comment (ericchiang) [inline db/refresh.go]: Why is verify outside the transaction?

**Comment by ericchiang** [comment]
small nit but other than that tested locally and lgtm

File: db/refresh.go (hunk 5 above)
Comment (rsoletob) [inline db/refresh.go]: Because if dex can't verify the token you save opening a transaction.

Comment (ericchiang) [inline db/refresh.go]: okay sure

**Comment by ericchiang** [comment]
lgtm!

**Pull request merged**